### ISSUE TITLE
Date/DateTime sort works correctly on question but not on dashboard

### ISSUE THREAD

**Issue by Wchua** [metabase/metabase#31205]
### Describe the bug

When i create a question (with query builder) returning a date or datetime column i can sort the column correctly on the editor.

![image](https://github.com/metabase/metabase/assets/8399465/04d9af8d-506a-4c81-b298-b32c28430c26)

But on the dashboard the sort appear to be random.

![image](https://github.com/metabase/metabase/assets/8399465/df642b75-70e9-4bd0-aedb-8bff0480cba5)

If i export the data (on dashboard) to csv
![image](https://github.com/metabase/metabase/assets/8399465/47a6e0e2-b2a1-4726-af10-892722044a66)

It's not sorting it correctly either.
NOTE: I have data ranging from 2017-12-05 00:00 to 2023-05-30 18:50 (YYYY-MM-DD HH:MM) with a total of 33 765 rows.

|Episodio|Data Realizacao|
|---------|------------------|
|30067071|2022-09-05T18:15:00|
|30067360|2022-09-05T19:00:00|
|30067361|2022-09-07T18:00:00|
|30068087|2022-09-19T17:30:00|
|30117934|2023-01-11T15:26:00|
|915657|2022-12-13T16:50:00|
|937460|2023-02-15T17:30:00|
|936692|2023-02-18T09:25:00|
|937140|2023-02-18T11:20:00|
|915687|2022-12-13T16:00:00|
|937221|2023-02-18T09:00:00|
|939675|2023-02-28T18:00:00|
|936079|2023-02-11T10:40:00|
|945962|2023-03-03T17:00:00|
|936061|2023-02-17T16:00:00|
|945946|2023-03-03T18:00:00|
|936084|2023-02-11T12:30:00|
|937462|2023-02-15T16:40:00|
|913643|2022-12-13T19:25:00|
|912836|2022-12-06T17:30:00|

### To Reproduce

1. Create a new question returning a date/datetime column
2. Add the question to a dashboard
3. Sort the column on the dashboard card
4. The column does not sort correctly

### Expected behavior

The sorted column should sort correctly by the ascending or descending order

### Information about your Metabase installation

```JSON
{
  "browser-info": {
    "language": "pt-PT",
    "platform": "Win32",
    "userAgent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/113.0.0.0 Safari/537.36",
    "vendor": "Google Inc."
  },
  "system-info": {
    "file.encoding": "Cp1252",
    "java.runtime.name": "OpenJDK Runtime Environment",
    "java.runtime.version": "11.0.12+7",
    "java.vendor": "Eclipse Foundation",
    "java.vendor.url": "https://adoptium.net/",
    "java.version": "11.0.12",
    "java.vm.name": "OpenJDK 64-Bit Server VM",
    "java.vm.version": "11.0.12+7",
    "os.name": "Windows Server 2016",
    "os.version": "10.0",
    "user.language": "en",
    "user.timezone": "Europe/London"
  },
  "metabase-info": {
    "databases": [
      "h2",
      "sqlserver"
    ],
    "hosting-env": "unknown",
    "application-database": "postgres",
    "application-database-details": {
      "database": {
        "name": "PostgreSQL",
        "version": "15.3"
      },
      "jdbc-driver": {
        "name": "PostgreSQL JDBC Driver",
        "version": "42.5.1"
      }
    },
    "run-mode": "prod",
    "version": {
      "date": "2023-04-28",
      "tag": "v0.46.2",
      "branch": "release-x.46.x",
      "hash": "8967c94"
    },
    "settings": {
      "report-timezone": null
    }
  }
}
```

### Severity

It leads to a poor data analysis in some business cases

**Comment by Wchua** [comment]
> The sorting at the query level is done via an ORDER BY in the database, while the dashboard one is pure frontend (which I think it might be treating dates as a string)

@paoliniluis thanks for the reply.
This is true even for the export of the complete data?

**Comment by kamilmielnik** [comment]
> The sorting at the query level is done via an ORDER BY in the database, while the dashboard one is pure frontend

Yes, sorting in dashboard is purely frontend and frontend does not have access to all data, but only to the first 2000 rows of the original question (see [screenshot](https://github.com/metabase/metabase/assets/6830683/0eb49ea6-bf01-481a-adb8-6b0ff6032a0a)). This is why reversing the sorting order in dashboard will give unexpected results, i.e. only slice of the data (but the data should still be in the correct order).

The only way around this is to perform sorting in the backend.

Setting the 2000 row limit broke this: #21005

> (which I think it might be treating dates as a string)

Those are ISO 8601 strings and they were designed with sorting in mind, so that should not be a problem.

----

> This is true even for the export of the complete data?

No, exporting complete data is performed in the backend.

However, downloaded data will be sorted according to the order set in the Question, not the order set in the Dashboard - we may have a confusing UX here, cc: @kdoh @vbenedetti 

----

I have not managed to reproduce the issue where data appears out of any order. AFAIK it will happen only if the original question query does not include sorting and data is out of order in the source database - I assume the last screenshot and table come from that case.

**Comment by WiNloSt** [comment]
Can confirm the behavior described by @kamilmielnik 

Sorting questions on a dashboard aka dash cards doesn't seem to modify their underlying questions while sorting an individual question on the question page will also modify their question and turn them ad-hoc. That's why downloading results on the question page will get the correct sorted result.

> I have not managed to reproduce the issue where data appears out of any order. AFAIK it will happen only if the original question query does not include sorting and data is out of order in the source database - I assume the last screenshot and table come from that case.

I believe that's the case as well. e.g. our Orders table doesn't have created at sorted, so if we put orders on a dashboard, sort by create at and download the dash card result, the created at will appear unsorted.

**Comment by calherries** [comment]
Product's guidance on this is to sort in the frontend if row_count < max_rows else sort in the backend by re-running the query with a new order by clause.

[slack context](https://metaboat.slack.com/archives/C057T1QTB3L/p1695884593071389)

**Comment by kamilmielnik** [comment]
Reproducible in 0.46.2.
Reproducible in `master` ([d35c779](https://github.com/metabase/metabase/commit/d35c779976ae6bd79369e9cc6bec8540908f37d2)).

**Comment by adam-james-v** [comment]
As of right now, Team DashViz will move this to proj-metabase-lib. In short, this project is probably better suited to metabase-lib rather than DashViz. It seems like a good candidate to either wait until the 'Make Tables Scrollable in Dashboards' project is underway and/or to use the existing 'Drill Through' approach that we already do for Questions.

---

@markbastian , @camsaul , and I briefly discussed the following:

If you're doing an order by in a dashcard, you'll need to run a query on the backend to get the correct rows anyway, so Mark and I had considered these options:

- use viz-settings to store a `:sort/order-by` key that is either:
  - used at `POST "/:dashboard-id/dashcard/:dashcard-id/card/:card-id/query"` to modify the query with the correct :order-by. or
  - used in a pre-processing middleware to do the same query adjustment. The query adjustment part was a bit unclear to both Mark and I, and feels like something related to metabase-lib.

At this stage we talked with Cam, who pointed out the following:
- On Questions we have 'Drill Through' behaviour that works in the way we want, so perhaps we should use that approach.
- This is not likely something we want in :viz-settings. If someone is just sorting a table as they're exploring a Dashboard, we don't want that sort to stick around necessarily, unless you explicitly hit 'save' somewhere. Besides, our viz-settings are a bit messy and a potential source for bugs, so adding to that might be detrimental.

Finally, @cdeweyx  linked this Project doc: https://www.notion.so/metabase/Let-creators-make-tables-scrollable-in-dashboards-instead-of-paginated-2524be9424c444f6be53899f79143d0c

And suggested that this particular bug may be solvable alongside the proper implementation of that project.

**Comment by lbrdnk** [comment]
Sorting by created at column on dashboard and in query builder yields different results on master [`d419e9a`](https://github.com/metabase/metabase/commit/d419e9acc4109c59878f4b6fc51a86e23f7dc051).

**Comment by ranquild** [comment]
It seems that FE-level sorting in dashboards is just broken by design. We cannot make it work. Some ideas how this can be handled:
- Remove FE sorting from dashcards at all since it's broken without doing anything else
- Remove FE sorting and instead add a dashboard specific drill to the QB that shows sorted data
- Allow to create an ad-hoc queries with `order-by` based on datacard queries from dashboards

I don't think that the issue belongs to QPD at the moment.

**Comment by guizaols** [comment]
Any news on that? I'm struggling with this issue as well. Before v0.50, everything was working fine, but now I can't downgrade, so my dashboards are really messed up.

**Comment by MonikaRiehn** [comment]
Same here, two questions show right order of months in their standalone linegraphs. But when I put them both in one card, order of dates is totally messed up. My stakeholder go crazy.
